In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv("traders_data.csv")

In [3]:
df

countryCode firstName lastName              traderId stockSymbol  \
0            TV   Vanessa    Green  MUVGtYHdeMdauAxzEJvt          FB   
1           NaN       NaN      NaN                   NaN          FB   
2            UZ    Brandi  Robbins  JTzVqzzIkFlrYUQbhnOR        AMZN   
3            BT   Allison    Davis  TzqyQTQjZGeLZuJqlLaQ        AMZN   
4            SZ     Blake   Martin  WVcMKILUHtBkEahrPLrG          FB   
..          ...       ...      ...                   ...         ...   
995          KZ     Vicki  Burgess  deTclcaRrihIYAUADJPN        AMZN   
996         NaN       NaN      NaN                   NaN        AMZN   
997          KZ     Vicki  Burgess  deTclcaRrihIYAUADJPN          FB   
998          CV     April    Floyd  stOzTFyGrgJGPgVPVTJQ        AMZN   
999          LA   Brandon    Berry  xIZVPeFzGRmMglPrQfhK          FB   

    stockName      tradeId        price  volume        tradeDatetime  
0    Facebook          NaN  1701.760000    58.0  2020-04-03 07:03:32  
1    Facebook  P5-3756422C          NaN    98.0  2020-01-13 08:29:53  
2      Amazon  X0-7401234c  1857.630560     NaN  2020-07-16 07:18:39  
3      Amazon  K3-3189326K  2078.440300    32.0  2020-07-07 13:59:03  
4    Facebook  U4-1266883G  1809.600000   212.0  2020-08-01 02:57:54  
..        ...          ...          ...     ...                  ...  
995    Amazon  j3-8758429Q  1554.962000   243.0  2020-03-12 19:42:37  
996    Amazon  K3-7867680F  2285.000000     4.0  2020-07-21 05:36:12  
997  Facebook  S8-7693720h  1935.760000   223.0  2020-02-05 23:28:16  
998    Amazon  z1-1234462u  2039.582452    52.0  2020-06-24 13:28:57  
999  Facebook  g7-2872947d  2332.000000    85.0  2020-03-25 04:07:42  

[1000 rows x 10 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   countryCode    795 non-null    object 
 1   firstName      795 non-null    object 
 2   lastName       795 non-null    object 
 3   traderId       795 non-null    object 
 4   stockSymbol    913 non-null    object 
 5   stockName      913 non-null    object 
 6   tradeId        909 non-null    object 
 7   price          898 non-null    float64
 8   volume         895 non-null    float64
 9   tradeDatetime  902 non-null    object 
dtypes: float64(2), object(8)
memory usage: 78.2+ KB


 
1) Dataset contains details of 1000 trading details in 10 features.  
2) Every feature have missing values.   
3) tradeDatetime data type is string that need to convert to datetime.    
4) Memory Usage: 78.2 KB   



## Date Preprocessing

In [5]:

# Updating datatype of 'tradeDatetime' from string to datetime
df['tradeDatetime'] = pd.to_datetime(df['tradeDatetime'])

# Removing the time stamp from the datatime 
df['tradeDate'] = df['tradeDatetime'].apply(lambda x: pd.to_datetime(x.date()))

df['tradeDate'] = pd.to_datetime(df['tradeDate'])
# traders_data['volume'] = traders_data['volume'].astype(int)



In [6]:
# To understand which column is categorical and which one is Continuous
df.nunique()

countryCode       19
firstName         20
lastName          20
traderId          20
stockSymbol        3
stockName          3
tradeId          909
price            879
volume           286
tradeDatetime    902
tradeDate        206
dtype: int64

In [7]:
df.describe()

price      volume
count   898.000000  895.000000
mean   1992.943372  148.887151
std     288.685224   86.763078
min    1501.000000    0.000000
25%    1751.314618   70.000000
50%    1988.310480  148.000000
75%    2235.975207  221.000000
max    2499.120000  300.000000

## Data Cleaning

Trade dataset contains trading details for Amazon, Facebook, Tesla, however our objective is to find suspicous data for Amazon for Feb and March month. So dataset need to remove the unnecessary data from traders dataset.

In [8]:
# Date range of traders dataset
print('Traders dataset start date : ',df['tradeDate'].min())
print('Traders dataset end date : ',df['tradeDate'].max())

Traders dataset start date :  2020-01-06 00:00:00
Traders dataset end date :  2020-08-01 00:00:00


In [9]:
df.isnull().sum()

countryCode      205
firstName        205
lastName         205
traderId         205
stockSymbol       87
stockName         87
tradeId           91
price            102
volume           105
tradeDatetime     98
tradeDate         98
dtype: int64

In [10]:
df.shape

(1000, 11)

In [11]:

# Define function to filter dataset
def filter_trader_data(df, company, start_date, end_date):
    '''
    Returns the filtered trader data based on company, start date and end date.

    Parameters:
        df (dataframe):The trader data which is to be filter.
        company (string): Company by which trader data need to filter.
        start_date (string) : Start date from where trader data require.
        end_date (string) : End date till trader data require.
        

    Returns:
        The trader dataframe which gets filtered.   
    '''
    return df[(df['stockSymbol']==company) & 
                            (df['tradeDate']>=start_date) & 
                            (df['tradeDatetime']<=end_date)].reset_index(drop=True)



In [12]:
traders_data = filter_trader_data(df,'AMZN','2020-02-01','2020-03-31')

In [13]:
traders_data.shape

(98, 11)

In [14]:
traders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 98 entries, 0 to 97
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   countryCode    74 non-null     object        
 1   firstName      74 non-null     object        
 2   lastName       74 non-null     object        
 3   traderId       74 non-null     object        
 4   stockSymbol    98 non-null     object        
 5   stockName      98 non-null     object        
 6   tradeId        92 non-null     object        
 7   price          87 non-null     float64       
 8   volume         85 non-null     float64       
 9   tradeDatetime  98 non-null     datetime64[ns]
 10  tradeDate      98 non-null     datetime64[ns]
dtypes: datetime64[ns](2), float64(2), object(7)
memory usage: 8.5+ KB


In [15]:
# Find missing data and their percentage
total = traders_data.isnull().sum().sort_values(ascending=False)
percent = ((traders_data.isnull().sum()/traders_data.isnull().count())*100).sort_values(ascending=False)
missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])
missing_data.head(20)

Total    Percent
traderId          24  24.489796
lastName          24  24.489796
firstName         24  24.489796
countryCode       24  24.489796
volume            13  13.265306
price             11  11.224490
tradeId            6   6.122449
tradeDate          0   0.000000
tradeDatetime      0   0.000000
stockName          0   0.000000
stockSymbol        0   0.000000

About 24.5% of data are missing from firstName, lastName and countryCode. It will be not effective if will impute these missing values with their mode values. So, deleting the missing values somewhat usually not the best solution. However, it can be useful when most values in a column are missing.

In [16]:

# Deleting the missing values
traders_data.dropna(axis=0, inplace=True)

# Reset the index of traders dataset
traders_data.reset_index(drop=True)

# Finally check the summary of trader dataset
traders_data.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 55 entries, 0 to 97
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   countryCode    55 non-null     object        
 1   firstName      55 non-null     object        
 2   lastName       55 non-null     object        
 3   traderId       55 non-null     object        
 4   stockSymbol    55 non-null     object        
 5   stockName      55 non-null     object        
 6   tradeId        55 non-null     object        
 7   price          55 non-null     float64       
 8   volume         55 non-null     float64       
 9   tradeDatetime  55 non-null     datetime64[ns]
 10  tradeDate      55 non-null     datetime64[ns]
dtypes: datetime64[ns](2), float64(2), object(7)
memory usage: 5.2+ KB


In [17]:
traders_data[['tradeDate','tradeId']].isnull().values.any()

False

In [18]:
Columns = ['countryCode', 'firstName', 'lastName', 'traderId', 'stockSymbol',
       'stockName', 'tradeId', 'price', 'volume', 'tradeDate']

# Selecting final columns
TradersDataClean=traders_data[Columns]

In [19]:
# TradersDataClean.nunique()
TradersDataClean.select_dtypes(include="O").columns

Index(['countryCode', 'firstName', 'lastName', 'traderId', 'stockSymbol',
       'stockName', 'tradeId'],
      dtype='object')

In [20]:
# Saving this final data for reference during deployment
TradersDataClean.to_pickle('TradersDataClean.pkl')


In [21]:
# TradersDataClean.nunique()
TradersDataClean.select_dtypes(include="O").columns

Index(['countryCode', 'firstName', 'lastName', 'traderId', 'stockSymbol',
       'stockName', 'tradeId'],
      dtype='object')

## Task
We want to find traders which made suspicious orders. To be a suspicious orders we consider the 
following rules:
- The trader has submitted an order above the high price/below the low price for a given day of a stock
- The trader has submitted an order in a date when the stock was not traded

If any suspicious orders are found, we want to do the following analysis:

- If more than one trader is found, rank by number of suspicious orders per trader.
- Try to find if there is a correlation between the nationality of the trader and the 
tendency to make suspicious orders 

In [28]:
def find_fraud_trader(df):
    df['suspicious'] = True
    df['suspicious'] = df.apply(lambda row:True if(pd.isnull(row['High']) | (row['price'] > row['High']) | (row['price'] < row['Low'])) else False, axis=1)
    
    return df

In [29]:
print("Number of trading : ",df.shape[0])
print("Number of suspicious trading : ",df[df['suspicious']==True].shape[0])
print("Number of genuine trading : ",df[df['suspicious']==False].shape[0])

Number of trading :  1000


KeyError: 'suspicious'